homeassistant/components/synology_dsm/common.py

```diff
@@ -0,0 +1,267 @@
+        )
+
+        # Reset not used API, information is not reset since it's used in device_info
+        if not self._with_security:
+            LOGGER.debug(
+                "Disable security api from being updated for '%s'",
+                self._entry.unique_id,
+            )
+            self.dsm.reset(self.security)
+            self.security = None
+
+        if not self._with_storage:
+            LOGGER.debug(
+                "Disable storage api from being updatedf or '%s'", self._entry.unique_id
+            )
+            self.dsm.reset(self.storage)
+            self.storage = None
+
+        if not self._with_system:
+            LOGGER.debug(
+                "Disable system api from being updated for '%s'", self._entry.unique_id
+            )
+            self.dsm.reset(self.system)
+            self.system = None
+
+        if not self._with_upgrade:
+            LOGGER.debug(
+                "Disable upgrade api from being updated for '%s'", self._entry.unique_id
+            )
+            self.dsm.reset(self.upgrade)
+            self.upgrade = None
+
+        if not self._with_utilisation:
+            LOGGER.debug(
+                "Disable utilisation api from being updated for '%s'",
+                self._entry.unique_id,
+            )
+            self.dsm.reset(self.utilisation)
+            self.utilisation = None
+
+    def _fetch_device_configuration(self) -> None:
+        """Fetch initial device config."""
+        self.information = self.dsm.information
+        self.network = self.dsm.network
+        self.network.update()
+
+        if self._with_security:
+            LOGGER.debug("Enable security api updates for '%s'", self._entry.unique_id)
+            self.security = self.dsm.security
+
+        if self._with_storage:
+            LOGGER.debug("Enable storage api updates for '%s'", self._entry.unique_id)
+            self.storage = self.dsm.storage
+
+        if self._with_upgrade:
+            LOGGER.debug("Enable upgrade api updates for '%s'", self._entry.unique_id)
+            self.upgrade = self.dsm.upgrade
+
+        if self._with_system:
+            LOGGER.debug("Enable system api updates for '%s'", self._entry.unique_id)
+            self.system = self.dsm.system
+
+        if self._with_utilisation:
+            LOGGER.debug(
+                "Enable utilisation api updates for '%s'", self._entry.unique_id
+            )
+            self.utilisation = self.dsm.utilisation
+
+        if self._with_surveillance_station:
+            LOGGER.debug(
+                "Enable surveillance_station api updates for '%s'",
+                self._entry.unique_id,
+            )
+            self.surveillance_station = self.dsm.surveillance_station
+
+    async def async_reboot(self) -> None:
+        """Reboot NAS."""
+        try:
+            await self._hass.async_add_executor_job(self.system.reboot)
+        except (SynologyDSMLoginFailedException, SynologyDSMRequestException) as err:
```

Comment:
honestly would like to address this in a new PR, since it needs a bit more consideration and thoughts, how to create a generic usable helper in this class `SynoApi` for interacting with the device 🤔

Reply:
Untested but something like this might do it

```python3
diff --git a/homeassistant/components/synology_dsm/common.py b/homeassistant/components/synology_dsm/common.py
index 8f7888cabe..0fdb4153e1 100644
--- a/homeassistant/components/synology_dsm/common.py
+++ b/homeassistant/components/synology_dsm/common.py
@@ -37,6 +37,19 @@ from .const import CONF_DEVICE_TOKEN
 LOGGER = logging.getLogger(__name__)
 
 
+def syno_api_wrapper(api_call: Callable):
+    """Synology api call decorator."""
+
+    async def inner(self) -> dict:
+        response = {}
+        try:
+            return await self._hass.async_add_executor_job(api_call)
+        except (SynologyDSMAPIErrorException, SynologyDSMRequestException) as err:
+            LOGGER.debug("Error from '%s': %s", self._entry.unique_id, err, exc_info=True)
+
+    return inner
+
+
 class SynoApi:
     """Class to interface with Synology DSM API."""
 
@@ -217,28 +230,20 @@ class SynoApi:
             )
             self.surveillance_station = self.dsm.surveillance_station
 
+    @syno_api_wrapper
     async def async_reboot(self) -> None:
         """Reboot NAS."""
-        try:
-            await self._hass.async_add_executor_job(self.system.reboot)
-        except (SynologyDSMLoginFailedException, SynologyDSMRequestException) as err:
-            LOGGER.debug("Exception:%s", err)
-            raise err
+        return self.system.reboot
 
+    @syno_api_wrapper
     async def async_shutdown(self) -> None:
         """Shutdown NAS."""
-        try:
-            await self._hass.async_add_executor_job(self.system.shutdown)
-        except (SynologyDSMLoginFailedException, SynologyDSMRequestException) as err:
-            LOGGER.debug("Exception:%s", err)
-            raise err
+        return self.system.shutdown
 
+    @syno_api_wrapper
     async def async_unload(self) -> None:
         """Stop interacting with the NAS and prepare for removal from hass."""
-        try:
-            await self._hass.async_add_executor_job(self.dsm.logout)
-        except (SynologyDSMAPIErrorException, SynologyDSMRequestException) as err:
-            LOGGER.debug("Logout from '%s' not possible:%s", self._entry.unique_id, err)
+        return self.dsm.logout
 
     async def async_update(self, now: timedelta | None = None) -> None:
         """Update function for updating API information."""
```

Reply:
this looks really awesome, but TBH I did not get it work, maybe also because of missing experience with creating decorators 🙈  but I did another try with 5a5a8d8